luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+	else
+		originalShieldDamages[weaponDefID] = tonumber(weaponDef.customParams.shield_damage)
+	end
+
+	if weaponDef.type == 'Flame' then
+		flameWeapons[weaponDefID] = weaponDef
+	end
+end
+
+for id, data in pairs(UnitDefs) do
+	if data.customParams.shield_radius then
+		shieldUnitDefs[id] = data
+	end
+end
+
+function gadget:MetaUnitAdded(unitID, unitDefID, unitTeam)
+	if shieldUnitsData[unitID] then
+		shieldUnitsData[unitID].team = unitTeam
+	end
+end
+
+function gadget:UnitFinished(unitID, unitDefID, unitTeam)
+	if shieldUnitDefs[unitDefID] then
+		shieldUnitsData[unitID] = {
+			isStatic = shieldUnitDefs[unitDefID].isStatic or false,        -- Used to prefer paralyze disable method for shields that don't move
+			team = unitTeam,                                               -- For future AOE damage mitigation
+			location = { 0, 0, 0 },                                        -- For future AOE damage mitigation
+			shieldEnabled = true,                                          -- Virtualized enabled/disabled state until engine equivalent is changed
+			shieldDamage = 0,                                              -- This stores the value of damages populated in ShieldPreDamaged(), then applied in GameFrame() all at once
+			shieldWeaponNumber = -1,                                       -- This is replaced with the real shieldWeaponNumber as soon as the shield is damaged
+			downtime = shieldUnitDefs[unitDefID].customParams.shield_downtime or 8, -- Defined in unitdef.customparams with a default fallback value
+			downtimeReset = 0,
+			shieldCoverageChecked = false,									-- Used to prevent expensive unit coverage checks being performed more than once per cycle
+			radius = shieldUnitDefs[unitDefID].customParams.shield_radius
+		}
+	end
+
+	-- Increases performance by reducing global unitDefID lookups
+	unitDefIDCache[unitID] = unitDefID
+end
+
+function gadget:UnitDestroyed(unitID)
+	shieldUnitsData[unitID] = nil
+	unitDefIDCache[unitID] = nil
+end
+
+function gadget:ProjectileCreated(proID, proOwnerID, weaponDefID)
+	-- Increases performance by reducing global projectileDefID lookups
+	projectileDefIDCache[proID] = weaponDefID
+end
+
+function gadget:ProjectileDestroyed(proID)
+	projectileDefIDCache[proID] = nil
+	projectileShieldHitCache[proID] = nil
+end
+
+local function triggerDowntime(unitID, weaponNum)
+	local shieldData = shieldUnitsData[unitID]
+
+	-- This method is used for mobile units with shields such as evocom cortex commander. This is far less efficient, but for smaller unit counts is OK.
+	spSetUnitShieldRechargeDelay(unitID, weaponNum, 120)
+
+	spSetUnitShieldState(unitID, weaponNum, false)
+	shieldData.downtimeReset = gameSeconds + shieldData.downtime
+	shieldData.shieldEnabled = false
+end
+
+local function removeCoveredUnits(shieldUnitID)
+	for unitID, shieldList in pairs(shieldedUnits) do
+		if shieldList[shieldUnitID] then
+			shieldList[shieldUnitID] = nil
+		end
+	end
```

Comment:
If a reverse lookup table is maintained, entries can be directly removed from both, without having to iterate the entire tables.

Reply:
Maintaining a table of `shieldUnitId[shieldedUnitId]` so depending on what you're doing, you do the lookup in the relevant table, rather than having to iterate half the time.